### ISSUE TITLE
How to generate speech directly from mel spectrograms of custom audio files

### ISSUE THREAD

**Issue by partha2409** [NVIDIA/waveglow#180]
I do not want to use the tacotron2, instead i just have the mel spectrogram of some random audio file. I have used the exact waveglow mel spectrogram settings of 80 mels, 1024 window length and 256 hop length, 22050 sampling rate.

I understand that a postnet is used in tacotron2, the output of which is fed as input to waveglow. Is it possible to just use the pre-trained waveglow directly on mel spectrograms with some pre-processing? Or is it possible to use just the  postnet part of tacotron2 passing mel spectrogram as inputs and then using that output to produce speech with waveglow?

TIA 
@rafaelvalle

**Comment by NoahDrisort** [comment]
> I do not want to use the tacotron2, instead i just have the mel spectrogram of some random audio file. I have used the exact waveglow mel spectrogram settings of 80 mels, 1024 window length and 256 hop length, 22050 sampling rate.
> 
> I understand that a postnet is used in tacotron2, the output of which is fed as input to waveglow. Is it possible to just use the pre-trained waveglow directly on mel spectrograms with some pre-processing? Or is it possible to use just the postnet part of tacotron2 passing mel spectrogram as inputs and then using that output to produce speech with waveglow?
> 
> TIA
> @rafaelvalle

I think you should consider line 137 and 138 in mel2samp.py file and change its process

**Comment by NoahDrisort** [comment]
I don't know what is value MAX_WAV_VALUE mean in mel2samp.py

**Comment by partha2409** [comment]
I think mel2samp.py is useful to run inference only if I train my own waveglow model. But I want to use the pretrained published model.

**Comment by partha2409** [comment]
Btw the MAX_WAV_VALUE is just used to normalise the audio signal value between -1 and +1.

**Comment by CookiePPP** [comment]
This what you're looking for?
```
import torch
from denoiser import Denoiser
from layers import TacotronSTFT, STFT
from hparams import create_hparams
import IPython.display as ipd
hparams = create_hparams()

# Load mels
from utils import load_wav_to_torch
stft = TacotronSTFT(hparams.filter_length, hparams.hop_length, hparams.win_length,
                    hparams.n_mel_channels, hparams.sampling_rate, hparams.mel_fmin,
                    hparams.mel_fmax)
def load_mel(path):
    audio, sampling_rate = load_wav_to_torch(path)
    if sampling_rate != stft.sampling_rate:
        raise ValueError("{} {} SR doesn't match target {} SR".format(
            sampling_rate, stft.sampling_rate))
    audio_norm = audio / hparams.max_wav_value
    audio_norm = audio_norm.unsqueeze(0)
    audio_norm = torch.autograd.Variable(audio_norm, requires_grad=False)
    melspec = stft.mel_spectrogram(audio_norm)
    return melspec
```
Load WaveGlow
```
waveglow_path = 'WAVEGLOW PATH HERE'
waveglow = torch.load(waveglow_path)['model']
waveglow.cuda().eval().half()
for k in waveglow.convinv:
    k.float()
denoiser = Denoiser(waveglow)
```
Load Spectrogram from Wav
```
# Run on .wav file
file_path = 'WAV PATH HERE'
mel_outputs_postnet = load_mel(file_path).cuda().half()

audio = []
with torch.no_grad():
    sigma_=0.5; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.7; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.8; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.9; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=1.0; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))

# Denoise Outputs
denoise_strength = 0.01
for i in audio:
    audio_denoised = denoiser(i, strength=denoise_strength)[:, 0]
    ipd.display(ipd.Audio(audio_denoised.cpu().numpy(), rate=hparams.sampling_rate))
```
Load Spectrogram from npy file
```
import numpy as np
file_path = 'NPY PATH HERE'
mel_outputs_postnet = torch.from_numpy(np.load(file_path)).unsqueeze(0).cuda().half()

audio = []
with torch.no_grad():
    sigma_=0.5; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.7; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.8; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=0.9; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))
    sigma_=1.0; audio.append(waveglow.infer(mel_outputs_postnet, sigma=sigma_))
    print("sigma = {}".format(sigma_)); ipd.display(ipd.Audio(audio[len(audio)-1][0].data.cpu().numpy(), rate=hparams.sampling_rate))

# Denoise Outputs
denoise_strength = 0.01
for i in audio:
    audio_denoised = denoiser(i, strength=denoise_strength)[:, 0]
    ipd.display(ipd.Audio(audio_denoised.cpu().numpy(), rate=hparams.sampling_rate))

**Comment by partha2409** [comment]
yes thanks :) Just had to combine mel2samp.py and inference.py

**Comment by YiPrograms** [comment]
Hi, I'm trying to do this too, but I still can't figure it out :(
mel2samp.py converts wav files to mel spectrograms that can be fed to the pre-trained model
But I have only mel spectrograms (no wav files), what should I do before feeding them to the model?
I've tried feeding them to the model directly and it only output noises, but they can be converted to audio with librosa's Griffin-Lim without problems

**Comment by Shruti-Raj-Vansh-Singh** [comment]
> yes thanks :) Just had to combine mel2samp.py and inference.py

Hi can you please share the code.
